### DISCUSSION TITLE
Tuple/array version of `evolve`

### DISCUSSION THREAD

**Discussion by alexamy** [remeda/remeda#943] [Q&A] [answered]
Hi! 
There is `evolve` for objects, but how to implement `evolve` for the array or tuple? Or maybe there is an option to use Remeda functions to get similar result (without `map`). 
For example:
```ts
import * as R from 'remeda';

const data = [
  ['12', '1 2 3'], 
  ['13', '4 5 6'],
];

// current approach with map
R.pipe(data, R.map(([answer, numbers]) => ([
  Number(answer),
  R.pipe(numbers, R.split(" "), R.map(Number)),
] as const));

// hypothetical example with more declarative style
R.pipe(data, R.evolve([
  Number, 
  R.piped(R.split(" "), R.map(Number)),
] as const));
```

**Comment by eranhirsch** [comment, marked as answer]
If I understand the example correctly, this is mainly useful when treating arrays as tuples (e.g., a fixed-length array). In those cases, evolve technically already works, albeit its API is a little less intuitive:

```ts
const result = R.evolve([1,2], { 0: R.add(1), 1: R.add(3) }); //=> { 0: 2, 1: 5 };
```

The output isn't an array, but it still adheres to the expected access pattern using indices:

```ts
result[0]; //=> 2;
result[1]: //=> 5;
```

The problem is typing! When building the type for evolve we decided not to support arrays to keep things simple. It's not impossible to type, but we didn't feel it was necessary then.

My only caveat is that if you use an array as a tuple and need a function like evolve, you might be better off redesigning the type to be a plain object instead. Often people will use a tuple to represent things like a Coordinate: `[x: number, y: number]` when it is just as valid to build it as `{ x: number, y: number }` (or even `Record<"x" | "y", number>`) which would be more ergonomic (e.g., `point.x` and not `point[0]`). If you **aren't** planning on **iterating** over the tuple or treating it like an array, it might not need to be an array. From the JS engine's perspective, an array is just an object with some extra functionality in its prototype; storing a tuple over an object doesn't take less or more memory.

**Comment by olsonpm** [comment]
When we're in control of our data structures, I agree a tuple can be better represented as an object.  However, often we are given a structure such as one from an api or another library.  In those cases, it's helpful to have utilities work with what we're given rather than add a transformation step.

My library's equivalent of evolve is [`update`](https://common-fp.org/docs/update) which takes an object, array (tuple) or Map.  The OP's example could look [something like this](https://common-fp.org/try-it#editorData=eyJjb2RlIjoiaW1wb3J0IHsgY29tcG9zZSwgbWFwVmFsdWVzLCBzcGxpdCwgdXBkYXRlIH0gZnJvbSAnY29tbW9uLWZwJ1xuXG5jb25zdCBkYXRhOiBBcnJheTxbc3RyaW5nLCBzdHJpbmddPiA9IFtcbiAgWycxMicsICcxIDIgMyddLCBcbiAgWycxMycsICc0IDUgNiddLFxuXTtcblxuY29uc3QgbWFwVG9OdW1iZXJzID0gY29tcG9zZShbc3BsaXQoJyAnKSwgbWFwVmFsdWVzKE51bWJlcildKVxuY29uc3QgdG9OdW1iZXJzID0gdXBkYXRlKFtcbiAgTnVtYmVyLFxuICBtYXBUb051bWJlcnMsXG5dKTxbc3RyaW5nLCBzdHJpbmddLCBbbnVtYmVyLCBudW1iZXJbXV0%2BXG5jb25zdCBnZXROdW1iZXJzID0gbWFwVmFsdWVzKHRvTnVtYmVycylcblxuY29uc3QgcmVzdWx0ID0gZ2V0TnVtYmVycyhkYXRhKVxuc2hvdyhyZXN1bHQpIiwibGFuZ3VhZ2UiOiJ0cyIsInZlcnNpb24iOjF9).

*naming variables is hard with contrived examples :)*

```ts
import { compose, mapValues, split, update } from 'common-fp'

const data: Array<[string, string]> = [
  ['12', '1 2 3'], 
  ['13', '4 5 6'],
];

const mapToNumbers = compose([split(' '), mapValues(Number)])
const toNumbers = update([
  Number,
  mapToNumbers,
])<[string, string], [number, number[]]>
const getNumbers = mapValues(toNumbers)

const result = getNumbers(data)
show(result)
```

typing for generic structure support does get hairy - and in the case of `update` I opted for devs to explicitly pass their types since mapping them caused for some incomprehensible tooltips.  Though imo typescript wasn't intended for functional programming so you're taking on the type battle regardless.
